content/developer/reference/guidelines.rst

```diff
@@ -999,17 +999,88 @@ More precise JS guidelines are detailed in the `github wiki  <https://github.com
+
+.. note::
+    The importance of an element in regard to this rule is left to the appreciation of the reviewer. If the structure of the template is hard to follow when reading just the first words of each line, it is probably a sign that some `name` attributes are missing on those lines.
+
+The name on the root element should follow the format `o-ComponentName`. And for children elements `o-ComponentName-partName` where:
+
+- `o-` is a must have Odoo prefix.
+- `ComponentName` is the PascalCase name of the current component/template.
+- `-` is a mandatory hyphen to separate the component name and the part name.
+- `partName` is the camelCase name of the child element referred by this `name` attribute.
+
+Attributes
+~~~~~~~~~~
+
+All the attributes of an element should be on the same line as the element itself. No character limit per line.
+
+.. note::
+    For extremely long lines, it is left to the appreciation of the reviewer to accept multi-line attributes within an element (but it is not mandatory).
+
+    In that case each attribute should be on its own line, with an indentation of 4 spaces relative to the element itself. The closing of the element (either `>` or `/>`) should be on its own line as well, with the same indentation as the corresponding opening character.
+
+The following order is applied for attributes:
+
+ #. `name`
+ #. `t-if`, `t-elif`, `t-else`
+ #. `t-foreach`, `t-as`
+ #. attributes specific to this type of element (eg. `type` for `<input>`, `src` for `<img>`)
+ #. `t-on-`
+ #. `class`
+ #. other generic attributes (eg. `title`, `aria-`)
+ #. `id`
+ #. OWL technical attributes (eg. `t-key`, `t-ref`)
+
+`t-att-` and `t-attf-` variations are ordered just after their static counter-part, in this order respectively.
+
+.. note:: `id` should not be used unless absolutely necessary (such as for making an anchor, or as the `for` target of a `label`), and then it should be uniquely generated.
+
+Style
+~~~~~
+
+Prefer using Bootstrap components and utilities over anything else, by applying their classes on the template. Custom style should only be used if absolutely necessary.
+
+Inside the `class` attribute itself, the following order should be respected:
+
+ #. Custom template class (only if necessary for adding custom style).
+ #. Bootstrap component classes (eg. `btn`, `card`).
+ #. Bootstrap layout classes (anything that changes the size or position of the element or its children, eg. `d-flex`, `container`, `mt-2`).
+ #. Bootstrap style classes (eg. `text-muted`, `bg-primary`).
+
+.. warning:: Before writing custom style you are expected to know all existing Bootstrap components, utilities, mixin and variables, as well as all of Odoo's extra components, utilities, mixin and variables.
+
+If necessary a `css` or `scss` file should be used, one per component/template. Inside that file, the style rules should be as flat as possible and target specific parts of the component/template.
```

Comment:
I disagree here actually. Doing so means that we have a flat structure on the css side and a tree structure on the html side. This is a mismatch, which is annoying.  The scss way to represent embedded structure makes it closer to the html structure, which is very useful.  Using a flat structure means that you need to edit two possibly very different place when you do stuff like position absolute/relative or displayflex (which are usually done on a parent/children and are cases where the parented relationship has some relevance).

Reply:
I wouldn't mind to leave this to the appreciation of the reviewer too, as long as I'm not forced to do it. But it's probably one of the only points where I feel there is no benefit to allow it and only drawbacks.
The mismatch of not having to follow the same structure is not an annoyance, it is a liberation. Much easier to know which rules apply on what, much easier to override them, much easier to move elements around.
If you are doing it correctly, you pretty much never have to update CSS when you change the template, unless you actually want to change the style.

I see your argument for position and flex, but for them we have to use bootstrap classes and not custom style (maybe not just for absolute, but then parent and child are not linked anyway).

Some interesting read in regard to this (and a lot of other) points: https://adamwathan.me/css-utility-classes-and-separation-of-concerns/

Reply:
I am unconvinced. I read the linked post, it is very interesting and I mostly agree with it. But the main point is that we should use utility classes, and I agree with that.

Reply:
also, agree on the file per component.